### PR TITLE
Refine ParallelGraph Execution

### PR DESCRIPTION
**Pull request by Yancey0623**

### CODE REVIEW

File: paddle/fluid/framework/details/all_reduce_deps_pass.h [hunk 1]
```diff
@@ -21,6 +21,8 @@ namespace paddle {
 namespace framework {
 namespace details {
 
+constexpr char kAllOpDescs[] = "all_op_descs";
```
Comment (panyx0718) [inline paddle/fluid/framework/details/all_reduce_deps_pass.h]: this also in memory_optimizer_helper.h, unify them?

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc [hunk 2]
```diff
@@ -226,7 +221,7 @@ std::unique_ptr<ir::Graph> MultiDevSSAGraphBuilderBase::ApplyImpl(
    * Only variables should be the leaves of graph.
    */
   AddOutputToLeafOps(&result);
-  result.Erase(kGraphOps);
+  // result.Erase(kGraphOps);
```
Comment (typhoonzero) [inline paddle/fluid/framework/details/multi_devices_graph_pass.cc]: Why not removing this?

File: paddle/fluid/framework/details/build_strategy.cc [hunk 3]
```diff
@@ -35,8 +35,8 @@ static inline bool SeqOnlyAllReduceOps(const BuildStrategy &strategy) {
   // Should fix the allreduce op order if scheduling
   // them in multiple threads or processes to avoid hang.
   return (!strategy.enable_sequential_execution_ &&
-          strategy.num_trainers_ > 1) ||
-         strategy.enable_parallel_graph_;
+          strategy.num_trainers_ > 1) &&
+         !strategy.enable_parallel_graph_;
```
Comment (typhoonzero) [inline paddle/fluid/framework/details/build_strategy.cc:41]: `parallel_graph` do not need to add deps?

File: paddle/fluid/framework/parallel_executor.cc [hunk 4]
```diff
@@ -319,18 +304,35 @@ ParallelExecutor::ParallelExecutor(
   }
 
   if (build_strategy.enable_parallel_graph_) {
+#ifdef PADDLE_WITH_CUDA
```
Comment (panyx0718) [inline paddle/fluid/framework/parallel_executor.cc:307]: put this block of code in ParallelSSAGraphExecutor?

File: paddle/fluid/framework/details/parallel_ssa_graph_executor.cc [hunk 5]
```diff
@@ -13,11 +13,70 @@
 // limitations under the License.
 
 #include "paddle/fluid/framework/details/parallel_ssa_graph_executor.h"
+#include "paddle/fluid/framework/ir/graph_helper.h"
 
 namespace paddle {
 namespace framework {
 namespace details {
 
+std::vector<std::unique_ptr<ir::Graph>> SeparateMultiDevicesGraph(
```
Comment (typhoonzero) [inline paddle/fluid/framework/details/parallel_ssa_graph_executor.cc]: This should be in a pass?

Comment (panyx0718) [inline paddle/fluid/framework/details/parallel_ssa_graph_executor.cc]: this function is not safe. Copying graph can drop graph attrs.
Maybe we shouldn't copy graph. Instead, we pass selected graph ops and vars to threaded graph executor. (change threaded_graph_executor interface.)

Comment (panyx0718) [reply paddle/fluid/framework/details/parallel_ssa_graph_executor.cc]: let's hide this in parallel graph executor for now. Splitting graph is not safely supported now. (add comment). We might want to remove this.

File: paddle/fluid/framework/details/build_strategy.cc (hunk 3 above)
Comment (Yancey0623) [reply paddle/fluid/framework/details/build_strategy.cc:41]: PG needs to execute this pass on each graph.

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc (hunk 2 above)
Comment (Yancey0623) [reply paddle/fluid/framework/details/multi_devices_graph_pass.cc]: It's easy to iter all ops, and I don't know why to delete this attr, I think it's useful to do sth. about the graph.

File: paddle/fluid/framework/details/all_reduce_deps_pass.h (hunk 1 above)
Comment (Yancey0623) [reply paddle/fluid/framework/details/all_reduce_deps_pass.h]: done, moved to graph.h

File: paddle/fluid/framework/parallel_executor.cc (hunk 4 above)
Comment (Yancey0623) [reply paddle/fluid/framework/parallel_executor.cc:307]: done.

File: paddle/fluid/framework/details/build_strategy.cc [hunk 6]
```diff
@@ -34,9 +34,11 @@ namespace details {
 static inline bool SeqOnlyAllReduceOps(const BuildStrategy &strategy) {
   // Should fix the allreduce op order if scheduling
   // them in multiple threads or processes to avoid hang.
+  // NOTE: ParallelExecutor would execute this pass on each graph, so
```
Comment (panyx0718) [inline paddle/fluid/framework/details/build_strategy.cc]: ParallelGraphExecutor

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc [hunk 7]
```diff
@@ -392,19 +386,33 @@ void MultiDevSSAGraphBuilderBase::CreateComputationalOp(ir::Graph *result,
...
+  };
+
+  if (!strategy_.enable_parallel_graph_)
+    op_handle = append_allreduce_op(local_scopes_, places_);
 
   for (size_t i = 0; i < places_.size(); ++i) {
-    auto &p = places_[i];
+    auto p = places_[i];
+    std::vector<Scope *> ss{local_scopes_[i]};
+    std::vector<platform::Place> ps{p};
```
Comment (panyx0718) [inline paddle/fluid/framework/details/multi_devices_graph_pass.cc]: put this in if?

File: paddle/fluid/framework/details/multi_devices_helper.h [hunk 8]
```diff
@@ -36,13 +36,20 @@ namespace details {
...
 // aux variables to represent dependency. Useful to resolve data hazard.
-typedef std::unordered_set<VarHandleBase*> GraphDepVars;
+typedef std::unordered_set<VarHandleBase *> GraphDepVars;
 const char kGraphDepVars[] = "dep_vars";
+
+// TODO(panyx0718): Clean this up as well.
+// all operators. NOTE that even we use a vector here, the operators is
+// unordered.
+typedef std::vector<OpHandleBase *> GraphOps;
+const char kGraphOps[] = "ops";
```
Comment (panyx0718) [inline paddle/fluid/framework/details/multi_devices_helper.h]: you don't need to expose these? you can get graph ops:
```
for (OpHandleBase *op : ir::FilterByNodeWrapper<OpHandleBase>(*graph))
```

File: paddle/fluid/framework/ir/graph.h [hunk 9]
```diff
@@ -26,6 +26,11 @@ limitations under the License. */
 
 namespace paddle {
 namespace framework {
+
+namespace details {
+constexpr char kAllOpDescs[] = "all_op_descs";
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph.h:34]: doc here that this is not recommended?

File: paddle/fluid/framework/details/build_strategy.cc (hunk 6 above)
Comment (Yancey0623) [reply paddle/fluid/framework/details/build_strategy.cc]: done.

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc (hunk 7 above)
Comment (Yancey0623) [reply paddle/fluid/framework/details/multi_devices_graph_pass.cc]: done.

File: paddle/fluid/framework/details/multi_devices_helper.h (hunk 8 above)
Comment (Yancey0623) [reply paddle/fluid/framework/details/multi_devices_helper.h]: done.

File: paddle/fluid/framework/ir/graph.h (hunk 9 above)
Comment (Yancey0623) [reply paddle/fluid/framework/ir/graph.h:34]: done.

**Review by panyx0718 — APPROVED** [review]

**Review by chengduoZH — APPROVED** [review]

**Comment by Yancey0623** [comment]
@panyx0718 @typhoonzero This PR can not work well with the `conditional block op`, will fix it at the next PR,  so please don't cherry-pick to 1.3 before fixing the bug.

**Pull request merged**